tests/aws/services/events/test_events.py

```diff
@@ -1000,3 +1084,259 @@ def _get_sqs_messages():
+    def test_put_list_with_prefix_describe_delete_rule(
+        self, bus_name, create_event_bus, put_rule, aws_client, snapshot
+    ):
+        if bus_name == "custom":
+            bus_name = f"bus-{short_uid()}"
+            snapshot.add_transformer(snapshot.transform.regex(bus_name, "<bus-name>"))
+            create_event_bus(Name=bus_name)
+
+        rule_name = f"test-rule-{short_uid()}"
+        snapshot.add_transformer(snapshot.transform.regex(rule_name, "<rule-name>"))
+        response = put_rule(
+            Name=rule_name,
+            EventPattern=json.dumps(TEST_EVENT_PATTERN),
+            EventBusName=bus_name,
+        )
+        snapshot.match("put-rule", response)
+
+        # NamePrefix required for default bus against AWS
+        response = aws_client.events.list_rules(NamePrefix=rule_name, EventBusName=bus_name)
+        snapshot.match("list-rules", response)
+
+        response = aws_client.events.describe_rule(Name=rule_name, EventBusName=bus_name)
+        snapshot.match("describe-rule", response)
+
+        response = aws_client.events.delete_rule(Name=rule_name, EventBusName=bus_name)
+        snapshot.match("delete-rule", response)
+
+        response = aws_client.events.list_rules(NamePrefix=rule_name, EventBusName=bus_name)
+        snapshot.match("list-rules-after-delete", response)
+
+    @markers.aws.validated
+    def test_put_multiple_rules_with_same_name(self, put_rule, aws_client, snapshot):
+        rule_name = f"test-rule-{short_uid()}"
+        snapshot.add_transformer(snapshot.transform.regex(rule_name, "<rule-name>"))
+
+        response = put_rule(
+            Name=rule_name,
+            EventPattern=json.dumps(TEST_EVENT_PATTERN),
+        )
+        snapshot.match("put-rule", response)
+
+        if is_aws_cloud():
+            time.sleep(10)
+
+        # put_rule updates the rule if it already exists
+        response = put_rule(
+            Name=rule_name,
+            EventPattern=json.dumps(TEST_EVENT_PATTERN),
+        )
+        snapshot.match("re-put-rule", response)
+
+        response = aws_client.events.list_rules(NamePrefix=rule_name)
+        snapshot.match("list-rules", response)
+
+    @markers.aws.validated
+    def test_list_rule_with_limit(self, put_rule, aws_client, snapshot):
+        snapshot.add_transformer(snapshot.transform.jsonpath("$..NextToken", "next_token"))
+        rule_name_prefix = "test-rule"
+        count = 6
+
+        for i in range(count):
+            rule_name = f"{rule_name_prefix}-{i}"
+            put_rule(
+                Name=rule_name,
+                EventPattern=json.dumps(TEST_EVENT_PATTERN),
+            )
+
+        response = aws_client.events.list_rules(Limit=int(count / 2))
+        snapshot.match("list-rules-limit", response)
+
+        response = aws_client.events.list_rules(
+            Limit=int(count / 2) + 2, NextToken=response["NextToken"]
+        )
+        snapshot.match("list-rules-limit-next-token", response)
```

Comment:
For safeguarding the tests against concurrent side effects and reducing the length of (unnecessary snapshots). Some of these list snapshots could be replaced with assertions or at least more selective snapshots (e.g., snapshot one item for each endpoint, assuming the others look similar).

Reply:
I think it is necessary to test that not only the correct number of rules are returned, but also the correct order of rules in each limited group. This would be possible to assert with both Arn and Name, but is it not preferable to compare the response from AWS and LS in as much detail as possible and therefore use snapshot even though some concurrency occurs?

Reply:
I guess there are ways around it. I think the method supports filtering by nameprefix or event bus? If we make those unique, we can avoid concurrent / leftover requests making the tests flake, but still snapshot it.